### PR TITLE
Circ 199: Alternate Loan Period With Pending Hold Request

### PR DESCRIPTION
**Pull request by jeremythuff**
Resolves [Circ 199](https://issues.folio.org/browse/CIRC-199): Alternate Loan Period With Pending Hold Request.

## Purpose

This PR introduces two changes in the initial due date applied by a loan policy. Namely, if the second request in queue is both a hold and has a status of OPEN_NOT_YET_FILLED, then the alternateRenewalFixedDueDateSchedules will be applied. With this PR this will be true for both Rolling and Fixed due date schedules.

## Approach

This has been accomplished by adding a RequestQueue to the calculateInitialDueDate method in LoanPolicy. With the RequestQueue it was possible to perform the necessary logic within the determineStrategy method of LoanPolicy.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java [hunk 1]
```diff
@@ -244,8 +248,13 @@ private DueDateStrategy determineStrategy(boolean isRenewal, DateTime systemDate
           getRenewalDueDateLimitSchedules(), this::errorForPolicy);
       }
       else {
-        return new RollingCheckOutDueDateStrategy(getId(), getName(),
-          getPeriod(loansPolicy), fixedDueDateSchedules, this::errorForPolicy);
+        FixedDueDateSchedules appliedDueDateSchedules = Objects.nonNull(alternateRenewalFixedDueDateSchedules)
+          && Objects.nonNull(requestQueue)
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java]: I think we may want to introduce an none Request Queue to handle circumstances like this.

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java [hunk 2]
```diff
@@ -264,6 +278,20 @@ else if(isFixed(loansPolicy)) {
     }
   }
 
+  private boolean isAlternateRenewal(RequestQueue requestQueue) {
+    Optional<Request> potentialRequest = requestQueue.getRequests().stream().skip(1).findFirst();
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java]: Why is the first request skipped?

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java [hunk 3]
```diff
@@ -264,6 +278,20 @@ else if(isFixed(loansPolicy)) {
...
+    boolean isAlternateRenewal = false;
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java]: Why do we need a temporary variable, rather than eagerly return true or false in the two places?

File: src/test/java/org/folio/circulation/domain/policy/FixedLoanPolicyCheckOutDueDateCalculationTests.java [hunk 4]
```diff
@@ -33,7 +33,7 @@ public void shouldUseOnlyScheduleAvailableWhenLoanDateFits() {
     Loan loan = loanFor(loanDate);
 
     final Result<DateTime> calculationResult = loanPolicy
-      .calculateInitialDueDate(loan);
+      .calculateInitialDueDate(loan, null);
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/domain/policy/FixedLoanPolicyCheckOutDueDateCalculationTests.java]: This null parameter only adds to my increasing sense that there are significant design issues with the calculation of a due date. That is outside the scope of this issue though.

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java (hunk 3 above)
Comment (jeremythuff) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java]: This is just a coding style preference. I find it harder to follow the flow of control when there are multiple exits from a method.

File: src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java (hunk 1 above)
Comment (jeremythuff) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java]: WOuld you want to do this wirthin the scope of this issue?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/LoanPolicy.java]: No, I'm just planting seeds (partly in my mind) for improvements

**Comment by marcjohnson-kint** [comment]
Closed whilst changes are made to better reflect the requirements in the story
